runtime/caml/platform.h

```diff
@@ -120,6 +92,231 @@ void caml_plat_broadcast(caml_plat_cond*);
+   a condition-variable fallback */
+typedef struct caml_plat_futex caml_plat_futex;
+
+typedef uint32_t caml_plat_futex_value;
+typedef _Atomic caml_plat_futex_value caml_plat_futex_word;
+
+/* Block while `futex` has the value `undesired`, until woken by `wake_all` */
+void caml_plat_futex_wait(caml_plat_futex* futex,
+                          caml_plat_futex_value undesired);
+/* Wake all threads `wait`ing on `futex` */
+void caml_plat_futex_wake_all(caml_plat_futex* futex);
+
+/* Define CAML_PLAT_FUTEX_FALLBACK to use the condition-variable
+   fallback, even if a futex implementation is available */
+#ifndef CAML_PLAT_FUTEX_FALLBACK
+#  if defined(_WIN32)                                   \
+  || (defined(__linux__) && defined(HAS_LINUX_FUTEX_H)) \
+  || defined(__FreeBSD__) || defined(__OpenBSD__)
+/*  These exist, but are untested
+     defined(__NetBSD__) || defined(__DragonFly__) */
+#  else
+#    define CAML_PLAT_FUTEX_FALLBACK
+#  endif
+#endif
+
+#if !defined(CAML_PLAT_FUTEX_FALLBACK)
+struct caml_plat_futex {
+  caml_plat_futex_word value;
+};
+#  define CAML_PLAT_FUTEX_INITIALIZER(value) { (value) }
+
+Caml_inline void caml_plat_futex_init(caml_plat_futex* ftx,
+                                      caml_plat_futex_value value) {
+  ftx->value = value;
+}
+Caml_inline void caml_plat_futex_free(caml_plat_futex* ftx) {
+  (void) ftx; /* noop */
+}
+#else
+struct caml_plat_futex {
+  caml_plat_futex_word value;
+  caml_plat_mutex mutex;
+  pthread_cond_t cond;
+};
+#  define CAML_PLAT_FUTEX_INITIALIZER(value) \
+  { value, CAML_PLAT_MUTEX_INITIALIZER, PTHREAD_COND_INITIALIZER }
+
+void caml_plat_futex_init(caml_plat_futex* ftx, caml_plat_futex_value value);
+void caml_plat_futex_free(caml_plat_futex*);
+#endif
+
+/* Barriers */
+
+/*
+ * A barrier that can be either single-sense (separate release and
+ * reset) or sense-reversing (unified release and reset).
+ *
+ * | Operation | `caml_plat_barrier_*` function      |
+ * |           |---------------+---------------------|
+ * |           | Single-sense  | Sense-reversing     |
+ * |-----------|---------------+---------------------|
+ * | Reset     | `reset`       | automatic at `flip` |
+ * | Arrive    | `arrive`      | `arrive`            |
+ * | Check     | `is_released` | `sense_has_flipped` |
+ * | Block     | `wait`        | `wait_sense`        |
+ * | Release   | `release`     | `flip`              |
+ *
+ * The lifecycle is as follows:
+ *
+ *      Reset (1 thread)          (other threads)
+ *              |                       |
+ *              +----------+------------+
+ *                         |
+ *                      Arrive (all threads)
+ *                         |
+ *                         | check arrival number
+ *              +----------+------------+
+ *              |                       |
+ *       Check or Block              Release
+ *     (non-final threads)         (final thread)
+ *              |                       |
+ */
```

Comment:
I think we should make extremely clear here that `Check`-ing or `Block`-ing on a barrier without prior calling `arrive` is a programming error and will not check or block anything. And repeat the warning in a comment preceding each of the `Check` and `Block` functions.

Reply:
I really don't think that's necessary. It would theoretically be perfectly fine to call `Check` or even `Block` by another thread that was merely monitoring and wasn't expected to arrive at all, and it should be evident that `Block`-ing without `Arrive`-ing first, where expected, would cause a guaranteed deadlock.